### DISCUSSION TITLE
Self output port

### DISCUSSION THREAD

**Discussion by fmontesi** [jolie/jolie#425] [Ideas (Features, Fixes, etc.)]
Recursion with a stack is usually achieved in Jolie by using self-calls to a `local` input port. I was wondering if we should introduce it as a default, that is, an output port called `self` that is always defined. Kinda like `this` for objects.

Defining `self` requires collecting all the operations defined for any input port of the service with location `local`. Say that these operations are accumulated in an interface called `SelfInterface` (or `LocalInterface`?). Then we could automatically define `self` as:
```jolie
outputPort self {
  location: "local"
  interfaces: SelfInterface
}
```

I don't know if `SelfInterface` should be available to the programmer: for now, the conservative thing to do is to keep it hidden.

What do you think?

**Comment by thesave** [comment]
The `self` port would be surely useful. What if we went the other way around and provided `self` as a special outputPort for which the user needs only to provide the interfaces? It seems quicker and more explicit than asking the user to create inputPorts at `local` locations and implicitly creating the `self` outputPort by gathering their interfaces.
This would mean adding automatically the `local` location to the `self` outputPort, while we'd keep hidden the complementary `local` inputPort on those interfaces.

What I have in mind is something like:

```jolie
type ListInt: void {
 item*: int
}

interface PublicInterface {
 requestResponse: sort( ListInt )( ListInt )
}

interface PrivateInterface {
 requestResponse: merge( ListInt )( ListInt )
}

service Main {

 inputPort in {
  location: "socket://..."
  protocol: http
  interfaces: PublicInterface
 }

 outputPort self {
  interfaces: PublicInterface, PrivateInterface
 }
 
 main {
  [ sort( i )( o ){
   ...
   { sort@self( i_1 )( o_1 ) | sort@self( i_2 )( o_2 ) }
   merge@self( { l << o_1, r << o_2 } )( o )
  } ]
  [ merge( i )( o ){
   ...
  } ]
 }

}
```

**Reply by fmontesi** [reply]
I like this direction!
The syntax you're using for the `self` output port is the same of "standard" output ports, so I'm not sure that this is clear. That looks like an output port without a location to me, but it will actually get populated with the location of the interpreter's local listener.

Maybe we can figure out a shorter/dedicated syntax and then explain how it expands?

**Reply by thesave** [reply]
Well, from the top of my head, one possibility is to introduce `symmetricPort`s, which desugar to a couple, `outputPort`-`inputPort`, where the `outputPort` keeps the name of the symmetric one while the `inputPort` adds a suffix, like `_in`.

```jolie
symmetricPort whatev {
 location: "local"
 interfaces: PrivateInterface, PublicInterface
}
```

A nice trait about this is that we are not reserving the name `self` (one can name it, litterally, whatever ;)).
Moreover, we are letting the user decide if they want to use the "local" location or something else. E.g., by setting the location to `"socket://"` and the protocol to e.g., `http` we'd be both exposing the service to outside calls and keeping the possibility to call ourselves (albeit in a quite-costly manner :D)

**Reply by fmontesi** [reply]
The problem with symmetric ports is that in many cases the protocol configuration of an input port and the dual output port do not coincide, and we don't have a good way of dualising configurations automatically. :) This is especially evident for `http`, where request parameters become response, etc.

This is further complicated by the fact that in your example (which is the one we care about), `"local"` would not be the right location for the output port. We need to populate that location dynamically.

I don't know how important these issues are.

**Reply by fmontesi** [reply]
Ah, but we could leave `protocol` out for now. It is unnecessary for local ports anyway.

How about a primitive that allows for defining an output port that has the same interface and location of an input port at a service?

I don't have a good idea for a keyword yet, but here's how it could work:

```jolie
service S {
   inputPort IP { location: "local" interfaces: I1, I2 }
   outputPort self from IP
}
```

We could do this in general for all service definitions, kinda like we do for `embed` already, e.g.,

```jolie
service Target { inputPort MyInput { ... } }
service S {
   outputPort target from Target::MyInput
}
```

We could also say that an output port that statically declares location `"local"` is automatically linked to the location of the enclosing service.

**Reply by thesave** [reply]
I see. 
Could it be a "problem" solved by semantic checks? 
For example, we can reject 
```jolie
symmetricPort whatev {
 location: "local"
 interfaces: PrivateInterface, PublicInterface
}
```
but we can accept 
```jolie
symmetricPort whatev {
 location: "local://A"
 interfaces: PrivateInterface, PublicInterface
}
```
because the location is statically fixed, right?
That check should be simple to implement for `self` ports and we can always extend it to other locations and protocols as soon as we figure out how to project/complement configurations between the two input/outputPorts.

**Reply by thesave** [reply]
> Ah, but we could leave `protocol` out for now. It is unnecessary for local ports anyway.
> 
> How about a primitive that allows for defining an output port that has the same interface and location of an input port at a service?
> 
> I don't have a good idea for a keyword yet, but here's how it could work:
> 
> ```jolie
> service S {
>    inputPort IP { location: "local" interfaces: I1, I2 }
>    outputPort self from IP
> }
> ```
> 
> We could do this in general for all service definitions, kinda like we do for `embed` already, e.g.,
> 
> ```jolie
> service Target { inputPort MyInput { ... } }
> service S {
>    outputPort target from Target::MyInput
> }
> ```
> 
> We could also say that an output port that statically declares location `"local"` is automatically linked to the location of the enclosing service.

That's also an interesting alternative; a bit more verbose than `symmetricPort`s but more powerful for sure :)

**Comment by fmontesi** [comment]
It'd be nice if whatever primitive we pick will be expressive enough to be
useful in the implementation of embed ... as ... clauses. At least
conceptually. 🙂
